tests/unit_tests/sparse_blas/include/common_sparse_reference.hpp

```diff
@@ -0,0 +1,230 @@
+                                 std::to_string(static_cast<char>(transpose_val)));
+    }
+    return std::make_tuple(iopa, jopa, opa);
+}
+
+template <typename fpType>
+auto dense_transpose_if_needed(const fpType *x, std::size_t outer_size, std::size_t inner_size,
+                               std::size_t ld, oneapi::mkl::transpose transpose_val) {
+    std::vector<fpType> opx;
+    if (transpose_val == oneapi::mkl::transpose::nontrans) {
+        opx.assign(x, x + outer_size * ld);
+    }
+    else {
+        opx.resize(outer_size * ld);
+        for (std::size_t i = 0; i < outer_size; ++i) {
+            for (std::size_t j = 0; j < inner_size; ++j) {
+                opx[i + j * ld] = x[i * ld + j];
+            }
+        }
+    }
+    return opx;
+}
+
+/// Convert the sparse matrix in the given format to a dense matrix A in row major layout applied with A_view.
+template <typename fpType, typename intType>
+std::vector<fpType> sparse_to_dense(sparse_matrix_format_t format, const intType *ia,
+                                    const intType *ja, const fpType *a, std::size_t a_nrows,
+                                    std::size_t a_ncols, std::size_t nnz, intType indexing,
+                                    oneapi::mkl::transpose transpose_val,
+                                    oneapi::mkl::sparse::matrix_view A_view) {
+    oneapi::mkl::sparse::matrix_descr type_view = A_view.type_view;
+    oneapi::mkl::uplo uplo_val = A_view.uplo_view;
+    const bool is_symmetric_or_hermitian_view =
+        type_view == oneapi::mkl::sparse::matrix_descr::symmetric ||
+        type_view == oneapi::mkl::sparse::matrix_descr::hermitian;
+    // Matrices are not conjugated if they are symmetric
+    const bool apply_conjugate =
+        !is_symmetric_or_hermitian_view && transpose_val == oneapi::mkl::transpose::conjtrans;
+    std::vector<fpType> dense_a(a_nrows * a_ncols, fpType(0));
+
+    auto write_to_dense_if_needed = [&](std::size_t a_idx, std::size_t row, std::size_t col) {
+        if ((type_view == oneapi::mkl::sparse::matrix_descr::triangular ||
+             is_symmetric_or_hermitian_view) &&
+            ((uplo_val == oneapi::mkl::uplo::lower && col > row) ||
+             (uplo_val == oneapi::mkl::uplo::upper && col < row))) {
+            // Read only the upper or lower part of the sparse matrix
+            return;
+        }
+        if (type_view == oneapi::mkl::sparse::matrix_descr::diagonal && col != row) {
+            // Read only the diagonal
+            return;
+        }
+        // Do not transpose symmetric matrices to simplify the propagation of the symmetric values
+        std::size_t dense_a_idx =
+            (!is_symmetric_or_hermitian_view && transpose_val != oneapi::mkl::transpose::nontrans)
+                ? col * a_nrows + row
+                : row * a_ncols + col;
+        fpType val = opVal(a[a_idx], apply_conjugate);
+        dense_a[dense_a_idx] = val;
```

Comment:
Just typing out cases to be verified. I'm mostly concerned about the `hermitian` case. I'll revisit this comment later.

|`transpose`|`matrix_descr`|`uplo`|`diag`|Dense output|Logic verified?|
|------------|---------------|-------|------|---------------|----------------|
|`nontrans`|`triangular`|`lower`|`nonunit`|$L + D + \mathbf{0}$| Yes |
|`trans`|`triangular`|`lower`|`nonunit`|$\mathbf{0} + D + L^T$| Yes |
|`conjtrans`|`triangular`|`lower`|`nonunit`|$\mathbf{0} + \overline{D} + \overline{L}^T$| Yes |
|`nontrans`|`triangular`|`upper`|`nonunit`|$\mathbf{0} + D + U$| Yes |
|`trans`|`triangular`|`upper`|`nonunit`|$U^T + D + \mathbf{0}$| Yes |
|`conjtrans`|`triangular`|`upper`|`nonunit`|$\overline{U}^T + \overline{D} + \mathbf{0}$| Yes |
|`nontrans`|`symmetric`|`lower`|`nonunit`|$L + D + L^T$| TBD |
|`trans`|`symmetric`|`lower`|`nonunit`|$L + D + L^T$| TBD |
|`conjtrans`|`symmetric`|`lower`|`nonunit`|$\overline{L} + \overline{D} + \overline{L}^T$| TBD |
|`nontrans`|`symmetric`|`upper`|`nonunit`|$U^T + D + U$| TBD |
|`trans`|`symmetric`|`upper`|`nonunit`|$U^T + D + U$| TBD |
|`conjtrans`|`symmetric`|`upper`|`nonunit`|$\overline{U}^T + \overline{D} + \overline{U}$| TBD |
|`nontrans`|`hermitian`|`lower`|`nonunit`|$L + D + \overline{L}^T$| TBD |
|`trans`|`hermitian`|`lower`|`nonunit`|$\overline{L} + D + L^T$| TBD |
|`conjtrans`|`hermitian`|`lower`|`nonunit`|$L + \overline{D} + \overline{L}^T$| TBD |
|`nontrans`|`hermitian`|`upper`|`nonunit`|$\overline{U}^T + D + U$| TBD |
|`trans`|`hermitian`|`upper`|`nonunit`|$U^T + D + \overline{U}$| TBD |
|`conjtrans`|`hermitian`|`upper`|`nonunit`|$\overline{U}^T + \overline{D} + U$| TBD |

Reply:
I thought some more about this and figured that it makes more sense to prevent the user from using `conjtrans` with symmetric or Hermitian matrices. No backend would support that and it is probably not useful. `spmv` is the only operation affected by this so I have disabled that in https://github.com/oneapi-src/oneMKL/pull/500/commits/ec728319293b8b179fc02fd0ffef494b541c5e2b. I've also clarified this in the spec.

Reply:
That's actually partially incorrect.
* `conjtrans` + `symmetric` (rows 9 and 12) are unique operations as opposed to `nontrans` (rows 7 and 8 which are same) or `trans` (rows 10 and 11 which are same).
* `trans` + `hermitian` (rows 14 and 17) are unique operations as opposed to `nontrans` (rows 13 and 15 which are same) or `trans` (rows 16 and 18 which are same).

Reply:
The set of duplications are:
* 8 duplicates 7
* 11 duplicates 10
* 15 duplicates 13
* 18 duplicates 16
So other than 8, 11, 15, 18, all the rest of the rows are valid operations.

Reply:
I don't understand whether this means you are suggesting a change in this PR?

Reply:
Yes, you mentioned you added restrictions on `symmetric`/`hermitian` in ec72831. All the combinations will be eventually supported although no backends currently support `hermitian` views, so you can throw a `not_implemented` exception for that. `symmetric` views are supported.

You also mentioned in your first comment:
> I've also clarified this in the spec.

Could you please point me to where this was?

Reply:
The related change in the spec was: https://github.com/uxlfoundation/oneAPI-spec/pull/542/commits/471a11d6366ed0d6c869c39b9d2756473a709036
I'm ok to revert this and throw unsupported "for now". Note that this means all backends with throw unsupported for some time.

Reply:
I have changed it to throwing an `unimplemented` exception in https://github.com/oneapi-src/oneMKL/pull/500/commits/c7a4420b1494eed0068b9cf8e38dd546cc404120
The symmetric + conjtrans issue that I described in my email "MKLCPU and MKLGPU mismatch for spmm using conjtrans and symmetric property" is still present with oneAPI 2024.2 so I am throwing unsupported for both symmetric and hermitian + conjtrans. I have reverted the spec change that I linked above.

Reply:
Sorry the link doesn't show it but it was part of this existing PR: https://github.com/uxlfoundation/oneAPI-spec/pull/542
I have already reverted that in that PR.